src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/DefaultRazorProjectHost.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using MonoDevelop.Projects;
+
+namespace Microsoft.VisualStudio.Mac.LanguageServices.Razor.ProjectSystem
+{
+    internal class DefaultRazorProjectHost : RazorProjectHostBase
+    {
+        private const string RazorLangVersionProperty = "RazorLangVersion";
+        private const string RazorDefaultConfigurationProperty = "RazorDefaultConfiguration";
+        private const string RazorExtensionItemType = "RazorExtension";
+        private const string RazorConfigurationItemType = "RazorConfiguration";
+        private const string RazorConfigurationItemTypeExtensionsProperty = "Extensions";
+
+        public DefaultRazorProjectHost(
+            DotNetProject project,
+            ForegroundDispatcher foregroundDispatcher,
+            ProjectSnapshotManagerBase projectSnapshotManager)
+            : base(project, foregroundDispatcher, projectSnapshotManager)
+        {
+        }
+
+        protected override async Task OnProjectChangedAsync()
+        {
+            ForegroundDispatcher.AssertBackgroundThread();
```

Comment:
Maybe I haven't seen the complete picture here but what guarantees this is called on a background thread?

Reply:
We invoke it ourselves via the foregorund dispatchers background task scheduler